src/mlpack/methods/ann/gan_impl.hpp

```diff
@@ -0,0 +1,255 @@
+  double res = discriminator.outputLayer.Forward(
+      std::move(boost::apply_visitor(
+          outputParameterVisitor, 
+          discriminator.network.back())), std::move(currentTarget));
+  noise.imbue( [&]() { return noiseFunction(randGen);} );
+  generator.Forward(std::move(noise));
+  discriminator.predictors.col(numFunctions) = boost::apply_visitor(
+      outputParameterVisitor, generator.network.back());
+  discriminator.Forward(std::move(discriminator.predictors.col(numFunctions)));
+  discriminator.responses(numFunctions) = 0;
+
+  currentTarget = discriminator.responses.unsafe_col(numFunctions);
+  res += discriminator.outputLayer.Forward(
+        std::move(boost::apply_visitor(
+            outputParameterVisitor, 
+            discriminator.network.back())), std::move(currentTarget));
+  return res;
+}
+
+template<typename Model, typename InitializationRuleType, typename NoiseType>
+void GAN<Model, InitializationRuleType, NoiseType>::
+Gradient(const arma::mat& /*parameters*/, const size_t i, arma::mat& gradient)
+{
+  if (!reset)
+    Reset();
+
+  if (gradient.is_empty())
+  {
+    if (parameter.is_empty())
+    {
+      Reset();
+    }
+    gradient = arma::zeros<arma::mat>(parameter.n_elem, 1);
+  }
+  else
+  {
+    gradient.zeros();
+  }
+
+  if (noiseGradientDiscriminator.is_empty())
+  {
+    noiseGradientDiscriminator = arma::zeros<arma::mat>(
+        gradientDiscriminator.n_elem, 1);
+  }
+  else
+  {
+    noiseGradientDiscriminator.zeros();
+  }
+
+  gradientGenerator = arma::mat(gradient.memptr(),
+      generator.Parameters().n_elem, 1, false, false);
+
+  gradientDiscriminator = arma::mat(gradient.memptr() + 
+      gradientGenerator.n_elem,
+      discriminator.Parameters().n_elem, 1, false, false);
+
+  // get the gradients of the discriminator
+  discriminator.Gradient(discriminator.parameter, i, gradientDiscriminator);
+
+  noise.imbue( [&]() { return noiseFunction(randGen);} );
+  generator.Forward(std::move(noise));
+  discriminator.predictors.col(numFunctions) = boost::apply_visitor(
+      outputParameterVisitor, generator.network.back());
+  discriminator.responses(numFunctions) = 0;
+
+  discriminator.Gradient(discriminator.parameter, numFunctions,
+      noiseGradientDiscriminator);
+
+  gradientDiscriminator += noiseGradientDiscriminator;
+
+  if (*currentBatch % generatorUpdateStep == 0 && *currentBatch != 0)
+  {
+    // Minimize log(1 - D(G(noise)))
+    // pass the error from discriminator to generator
+    generator.error = boost::apply_visitor(deltaVisitor,
+        discriminator.network[1]);
+    generator.currentInput = noise;
+    generator.Backward();
+    generator.ResetGradients(gradientGenerator);
+    // set the current input requrired for gradient computation
+    generator.Gradient();
+
+    double multiplier = 1.0;
+
+    gradientGenerator = -gradientGenerator;
+    gradientGenerator *= multiplier;
+
+    if (counter % batchSize == 0)
```

Comment:
I suggest to restore batch handling here (see the comment above) until the rest of optimizers get refactored.